rfcs/0032-automated-validation.md

```diff
@@ -0,0 +1,219 @@
+
+#### Alerts
+[detailed-design-architecture-alerts]: #detailed-design-architecture-alerts
+
+When validations fail, alerts can be triggered. In theory, there can be multiple backends for how alerts are surfaced to the team. As an initial scope for the project, I think that focusing on discord alerts to an internal channel which is monitored by the testnet guardian and other engineers should suffice for now. One important note here is that this service should be *loud*, but not *noisy*. This means that the service should send a notification somewhere people see it, but it should contain some logic to help limit the noise level of these alerts. We don't want to have this alerting system spam too much and make uneccessary noise, because that will just lead people to ignore it like we do much of the noise from failing CI tests. As such, a basic validation error rate limiter and timeout logic seems to be a must.
+
+### Language Choice
+[detailed-design-language-choice]: #detailed-design-language-choice
+
+Language choice is a key component for the automated validation service design which has large ramifications on it's implementation and maintenance costs. In particular, we want to optimize for the maintenance costs of the automated validation service. There are 2 important aspects to consider when optimizing for the maintenance cost of this service: scalability and reusability. Scalability, in the context of this service, directly relates to how concurrent this system can be, since this service will be monitoring networks in the size order of 100s of nodes. Reusability relates to our ability to configure and resuse abstractions for the individual components involved in the service (resources, providers, statistics, validations). In addition to these, we also would like to choose a language which either engineers already know, or a language which would be easy for engineers to pick up without much learning overhead.
+
+#### Languages Not Chosen
+
+With the requirement that this system be highly concurrent (so that we don't need to spend a lot of engineering effort trying to optimize this tool), that immediately hurts the case for the main language we use currently, OCaml. OCaml comes with great benefits, such as it already being our standard language on the team and it having a good static type system, but the single threaded concurrency story in OCaml is poor, and the solutions and tooling around them (rpc\_parallel) are bulky to use and quite difficult to iterate on. As such, OCaml seems like the wrong choice for this system as we would spend a lot of engineering effort on just the architecture for this system (and would likely need to revisit that as we intend to scale this system up to more and more validations and larger network sizes).
```

Comment:
Well, there's good-ish old system threads in OCaml, which you get by linking a library. It doesn't map to multiple cores, though. It's much lighter-weight than `rpc_parallel`.

Reply:
I assume you are referring to https://caml.inria.fr/pub/docs/manual-ocaml/libref/Thread.html ? It's certainly lighter-weight compared to `rpc_parallel`, but that comes at the cost of usability since we would need to write our own tooling for thread management and architect our own concurrent system, which seems like a lot of work.